artichoke-backend/src/extn/core/kernel/integer.rs

```diff
@@ -0,0 +1,177 @@
+            Ok(Self {
+                arg,
+                radix,
+                raise_exception,
+            })
+        } else {
+            Err(Error::NoImplicitConversionToString(
+                value.ruby_type().class_name().into(),
+                raise_exception,
+            ))
+        }
+    }
+}
+
+pub fn method(interp: &Artichoke, args: &Args) -> Result<Value, Error> {
+    let arg = args.arg.as_str();
+    let radix = args.radix;
+    let raise_exception = args.raise_exception;
+
+    let mut string = String::new();
+
+    // If have mutliple consecutive embedded underscores, argument error!
+    let multi_underscore_re = Regex::new(r"__+").unwrap();
+    if arg.starts_with('_') || arg.ends_with('_') || multi_underscore_re.is_match(arg) {
+        return Err(Error::InvalidValue(arg.into(), raise_exception));
+    }
+
+    // If have multiple consecutive leading/trailing signs, argument error!
+    let multi_sign_re = Regex::new(r"\+\++|\-\-+").unwrap();
+    if multi_sign_re.is_match(arg) {
+        return Err(Error::InvalidValue(arg.into(), raise_exception));
+    }
+
+    // Remove embedded underscore `_`
+    // because they represent error in `from_str_radix` & `from_str`
+    let arg = arg.replace('_', "");
+
+    // Remove leading and trailing white space,
+    // because they represent error in `from_str_radix` & `from_str`.
+    let arg = arg.trim();
+    let arg = if arg.starts_with('-') {
+        string.push('-');
+        &arg[1..]
+    } else if arg.starts_with('+') {
+        string.push('+');
+        &arg[1..]
+    } else {
+        arg
+    };
+
+    // if `arg` is null byte, raise `ArgumentError`
+    if arg.contains('\0') {
+        return Err(Error::ContainsNullByte(raise_exception));
+    }
+
+    let mut parsed_radix = None;
+    if arg.starts_with('0') && arg.len() > 2 {
+        match &arg[0..2] {
+            "0b" | "0B" => {
+                string.push_str(&arg[2..]);
+                parsed_radix = Some(2);
+            }
+            "0o" | "0O" => {
+                string.push_str(&arg[2..]);
+                parsed_radix = Some(8);
+            }
+            "0d" | "0D" => {
+                string.push_str(&arg[2..]);
+                parsed_radix = Some(10);
+            }
+            "0x" | "0X" => {
+                string.push_str(&arg[2..]);
+                parsed_radix = Some(16);
+            }
+            prefix if &prefix[0..1] == "0" => {
+                string.push_str(&arg[1..]);
+                parsed_radix = Some(8);
+            }
+            _ => {}
+        };
+    } else {
+        string.push_str(arg);
+    }
+
+    match (radix, parsed_radix) {
+        (Some(radix), Some(parsed_radix)) => {
+            if radix != parsed_radix {
+                return Err(Error::InvalidValue(string, raise_exception));
+            }
+            if let Ok(v) = i64::from_str_radix(string.as_str(), radix as u32) {
```

Comment:
`radix as u32` is an unsafe cast. Not all values that can be represented as an `i64` (including negative ones!) can be represented in a u32.

Instead we should use [`TryFrom`](https://doc.rust-lang.org/std/convert/trait.TryFrom.html) to return an error instead of doing something unexpected.

Also, it looks like Ruby takes the absolute value of the passed radix:

```
[2.6.3] > Integer('23', -10, exception: nil)
=> 23
[2.6.3] > Integer('23', -100, exception: nil)
Traceback (most recent call last):
        6: from /usr/local/var/rbenv/versions/2.6.3/bin/irb:23:in `<main>'
        5: from /usr/local/var/rbenv/versions/2.6.3/bin/irb:23:in `load'
        4: from /usr/local/var/rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        3: from (irb):10
        2: from (irb):10:in `rescue in irb_binding'
        1: from (irb):10:in `Integer'
ArgumentError (invalid radix 100)
```

`-1` radix is treated as `10`?

```
[2.6.3] > Integer('23', -1, exception: nil)
=> 23
```